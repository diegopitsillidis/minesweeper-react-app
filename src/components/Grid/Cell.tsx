import React, { FC } from 'react';
import styled from '@emotion/styled';

import {Cell as CellType, CellState } from '@/helpers/Field'

export interface CellProps {
    /**
     * 
     */
    children: CellType;
};

export const Cell: FC<CellProps> = ({ children }) => {
    switch (children) {
        case CellState.empty:
            return <RevealedFrame  />;
        case CellState.bomb:
            return (
                <BombFrame>
                    <Bomb/>
                </BombFrame>
            );
        case CellState.flag:
            return (
                <ClosedFrame>
                    <Flag/>
                </ClosedFrame>
            );
        case CellState.weakFlag:
            return (
                <ClosedFrame>
                    <WeakFlag />
                </ClosedFrame>
            );
        case CellState.hidden:
            return <ClosedFrame  />;
        default:
            return <RevealedFrame>{children}</RevealedFrame>;
    }
};

const ClosedFrame = styled.div`
    display: flex;
    align-items: center;
    justify-content: center;
    user-select: none;
    cursor: pointer;
    width: 1.8vw;
    height: 1.8vw;
    background-color: #d1d1d1;
    border: 0.6vh solid transparent;
    border-color: white #9e9e9e #9e9e9e white;
    &:hover {
        filter: brightness(1.1);
    }
`;

const transparent = 'rgba(0,0,0,0)';
const colors: { [key in CellType]: string } = {
    0: '#000',
    1: '#2a48ec',
    2: '#2bb13d',
    3: '#2bb13d',
    4: '#ec6561',
    5: '#a6070f',
    6: '#e400af',
    7: '#906a02',
    8: '#fa0707',
    9: transparent,
    10: transparent,
    11: transparent,
    12: transparent,
};

const RevealedFrame = styled(ClosedFrame)`
    border-color: #dddddd;
    cursor: default;
    color: ${({ children }) => colors[children as CellType] ?? transparent};
    &:hover {
        filter: brightness(1);
    }
`;

const Bomb = styled.div`
    border-radius: 50%;
    width: 1vw;
    height: 1vw;
    background-color: #333;
`;

const BombFrame = styled(RevealedFrame)`
    background-color: #ec433c;
`;

const Flag = styled.div`
    width: 0px;
    height: 0px;
    border-top: 0.5vw solid transparent;
    border-bottom: 0.5vw solid transparent;
    border-left: 0.5vw solid #ec433c;
`;

const WeakFlag = styled(Flag)`
    border-left: 0.5vw solid #f19996
`;