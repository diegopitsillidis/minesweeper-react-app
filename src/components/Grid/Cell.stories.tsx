import React from 'react';
import { StoryFn, Meta} from '@storybook/react'

import { Cell, CellProps } from './Cell';
import { Coords } from '@/core/Field';


export default {
    title: 'Grid/Cell',
    component: Cell,
    args: {
        coords: [1,1]
    }
} as Meta;

const Template: StoryFn<CellProps> = (args) => <Cell {...args} />;

export const CellClosed = Template.bind({});
CellClosed.args = {
    children: 10,
    onClick: (coords: Coords) => console.log('Cell clicked', coords),
    onContextMenu: (coords: Coords) => console.log('Context menu', coords),
}

export const CellIsEmpty = Template.bind({});
CellIsEmpty.args = {
    children: 0,
}

export const CellIsBomb = Template.bind({});
CellIsBomb.args = {
    children: 9,
}

export const CellIsFlag = Template.bind({});
CellIsFlag.args = {
    children: 11,
}

export const CellIsWeakFlag = Template.bind({});
CellIsWeakFlag.args = {
    children: 12,
}

export const CellWith1 = Template.bind({});
CellWith1.args = {
    children: 1,
}

export const CellWith3 = Template.bind({});
CellWith3.args = {
    children: 3,
}

export const CellWith5 = Template.bind({});
CellWith5.args = {
    children: 5,
}

export const CellWith8 = Template.bind({});
CellWith8.args = {
    children: 8,
}