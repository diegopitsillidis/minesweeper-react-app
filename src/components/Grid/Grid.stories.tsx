import React from 'react';
import { StoryFn, Meta} from '@storybook/react'
import { Grid, GridProps } from './Grid';


export default {
    title: 'Grid/Grid',
    component: Grid,
} as Meta;

const Template: StoryFn<GridProps> =  (args) => <Grid {...args} />

export const GridExample = Template.bind({});
GridExample.args = {
    children: [
        [9,1, 0],
        [1,1, 1],
        [10,10, 1],
    ],
    onClick(coords) {
        console.log(coords)
    },
}