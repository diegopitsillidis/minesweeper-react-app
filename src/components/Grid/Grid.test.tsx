import React from 'react';
import { render } from "@testing-library/react";
import { Grid, GridProps } from './Grid';

it('Grid renders correctly', () => {
    const args = {
        children: [
            [9,1, 0],
            [1,1, 1],
            [10,10, 1],
        ],
        onClick(coords) {
            console.log(coords)
        },
    } as GridProps;

    const { asFragment } = render(<Grid {...args} />);

    expect(asFragment()).toMatchSnapshot();
});