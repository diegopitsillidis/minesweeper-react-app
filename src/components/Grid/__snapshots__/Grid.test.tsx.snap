// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid renders correctly 1`] = `
<DocumentFragment>
  <div
    class="css-1qaavfo"
    size="3"
  >
    <div
      class="css-4dvszr"
      data-testid="9_0,0"
      role="cell"
    >
      <div
        class="css-1dc6pdv"
      />
    </div>
    <div
      class="css-u2nlwm"
      data-testid="1_0,1"
      role="cell"
    >
      1
    </div>
    <div
      class="css-1lpo8q3"
      data-testid="0_0,2"
      role="cell"
    />
    <div
      class="css-u2nlwm"
      data-testid="1_1,0"
      role="cell"
    >
      1
    </div>
    <div
      class="css-u2nlwm"
      data-testid="1_1,1"
      role="cell"
    >
      1
    </div>
    <div
      class="css-u2nlwm"
      data-testid="1_1,2"
      role="cell"
    >
      1
    </div>
    <div
      class="css-z6puck"
      data-testid="10_2,0"
      role="cell"
    />
    <div
      class="css-z6puck"
      data-testid="10_2,1"
      role="cell"
    />
    <div
      class="css-u2nlwm"
      data-testid="1_2,2"
      role="cell"
    >
      1
    </div>
  </div>
</DocumentFragment>
`;
